devices/src/virtio/vsock/packet.rs

```diff
@@ -0,0 +1,326 @@
+
+impl VsockPacket {
+    /// Create the packet wrapper from a TX virtq chain head.
+    ///
+    /// The chain head is expected to hold valid packet header data. A following packet buffer
+    /// descriptor can optionally end the chain. Bounds and pointer checks are performed when
+    /// creating the wrapper.
+    ///
+    pub fn from_tx_virtq_head(head: &DescriptorChain) -> Result<Self> {
+        // All buffers in the TX queue must be readable.
+        //
+        if head.is_write_only() {
+            return Err(VsockError::UnreadableDescriptor);
+        }
+
+        let hdr = HdrWrapper::from_virtq_desc(head)?;
+
+        // Reject weirdly-sized packets.
+        //
+        if hdr.len > defs::MAX_PKT_BUF_SIZE as u32 {
+            return Err(VsockError::InvalidPktLen(hdr.len));
+        }
+
+        // Don't bother to look for the data descriptor, if the header says there's no data.
+        //
+        if hdr.len == 0 {
+            return Ok(Self { hdr, buf: None });
+        }
+
+        let buf_desc = head.next_descriptor().ok_or(VsockError::BufDescMissing)?;
+
+        // All TX buffers must be readable.
+        //
+        if buf_desc.is_write_only() {
+            return Err(VsockError::UnreadableDescriptor);
+        }
+
+        // The data descriptor should be large enough to hold the data length indicated by the header.
+        //
+        if buf_desc.len < hdr.len {
+            return Err(VsockError::BufDescTooSmall);
+        }
+
+        Ok(Self {
+            hdr,
+            buf: Some(BufWrapper::from_virtq_desc(&buf_desc)?),
+        })
+    }
+
+    /// Create the packet wrapper from an RX virtq chain head.
+    ///
+    /// There must be two descriptors in the chain, both writable: a header descriptor and a data
+    /// descriptor. Bounds and pointer checks are performed when creating the wrapper.
+    ///
+    pub fn from_rx_virtq_head(head: &DescriptorChain) -> Result<Self> {
+        // All RX buffers must be writable.
+        //
+        if !head.is_write_only() {
+            return Err(VsockError::UnwritableDescriptor);
+        }
+
+        let hdr = HdrWrapper::from_virtq_desc(head)?;
+
+        let buf_desc = head.next_descriptor().ok_or(VsockError::BufDescMissing)?;
+        if !buf_desc.is_write_only() {
+            return Err(VsockError::UnwritableDescriptor);
+        }
+
+        Ok(Self {
+            hdr,
+            buf: Some(BufWrapper::from_virtq_desc(&buf_desc)?),
+        })
+    }
+
+    /// Builder-like method for setting `hdr.len` (the packet data length).
+    ///
+    pub fn set_len(&mut self, len: u32) -> &mut Self {
+        self.hdr.len = len;
+        self
+    }
+
+    /// Builder-like method for setting the `hdr.op` (the packet operation type).
+    ///
+    pub fn set_op(&mut self, op: u16) -> &mut Self {
+        self.hdr.op = op;
+        self
+    }
+
+    /// Builder-like method for setting a header flag.
+    ///
+    pub fn set_flag(&mut self, flag: u32) -> &mut Self {
```

Comment:
There's a `set_flags` to set the whole flags dword, and a `set_flag` to set a specific flag (OR it with whatever was there). It made a bit of sense to me to have both, but I can see that it may bring some confusion along as well. I don't have a strong opinion here.